DOC-8558: Additional discussion of deletion filters

tonyjhillman opened the pull request.

tonyjhillman committed:
DOC-8558: Additional discussion of deletion filters

tonyjhillman committed:
DOC-8558: Additional wording

tonyjhillman committed:
DOC-8558: Additional text on deletion filters and migration

tonyjhillman committed:
DOC-8558: Added deletion-filter related steps to migration procedure

tonyjhillman committed:
DOC-8558: Added cli and rest notes on deletion filters

tonyjhillman committed:
DOC-8558: Modified subsection title and text

nelio2k commented on modules/learn/pages/clusters-and-availability/xdcr-filtering.adoc:
```diff
@@ -163,3 +170,72 @@ image::xdcr/filter-replication-diagram-7.png[,720,align=left]
+=== Configuring Deletion-Filters to Protect Data
+
+Appropriate deletion-filter settings protect data.
+However, in certain circumstances, inappropriate deletion-filter settings may cause _loss_ of data.
```
This example talks about the loss of data. There is also situations where an inappropriate filter setting may cause accidental leak of stale data.

For example, if the user turns on the deletion filter in a replication, and the situation is where source bucket contains documents that contains sensitive information. If a source document was deleted for compliance purposes, then the target document could become stale and remain, even if were not meant to exist anymore.

It may help to create an example illustrating the opposite situation, covering both scenarios (if space or context suffices)

tonyjhillman replied:
I've added a similar subsection, which hopefully addresses this.

nelio2k commented on modules/learn/pages/clusters-and-availability/xdcr-filtering.adoc:
```diff
@@ -163,3 +170,72 @@ image::xdcr/filter-replication-diagram-7.png[,720,align=left]
+However, since document-deletions are replicated by default, replication _2_ has deleted documents of type _A_ from the target; and the actual outcome is therefore that only documents of type _B_ exist on the target; with documents of type _A_ existing on neither source nor target.
+
+To avoid this outcome, replication _2_ could be created with deletion filters configured to prevent the replication of deletions: the prior deletions of documents of type _A_ from the source would thereby _not_ be replicated to the target.
+Note, however, that this would also prevent the replication of future source-deletions of type _B_ documents: therefore, the creation of an entirely new source collection for type _B_ documents might be required, by means of _migration_, prior to the creation of replication _2_.
```
I find this line confusing. I'm not sure what migration has to do specifically with the concept of deletion being talked about here.
The suggestion here may cause more questions than answers since the deletion concept applies not only to  the scope of collections migration.

tonyjhillman replied:
Perhaps I've misunderstood. Is it just the last sentence that is problematic? ("Note, however," ... etc)

tonyjhillman replied:
Okay, I've removed the line with the reference to migration.

nelio2k commented on modules/learn/pages/clusters-and-availability/xdcr-filtering.adoc:
```diff
@@ -163,3 +170,72 @@ image::xdcr/filter-replication-diagram-7.png[,720,align=left]
+
+* Keep deletion filters at their default setting, to permit the replication of deletions.
+When the migration is judged to have completed, delete the migration _prior to_ the deletion of any source data.
+Then, once the migration is deleted, delete source data as appropriate.
```
Additional note as a reminder - if the replication is re-created after this point without a deletion filter, and the tombstones from the deletes have not been purged, then the deletes will be replicated to the target bucket.

tonyjhillman replied:
I've added a note to this effect.

nelio2k commented on modules/manage/pages/manage-xdcr/filter-xdcr-replication.adoc:
```diff
@@ -172,6 +172,9 @@ Conversely, if this option is not checked, the TTL _is_ made part of the replica
 
 For information on TTL and expiration, see xref:learn:buckets-memory-and-storage/expiration.adoc[Expiration].
 
+Note that the replication of deletions is _not_ prevented by the specifying of a filter, formed with regular and other filtering expressions: to ensure that document-deletions are _not_ replicated, the appropriate deletion-filter checkboxes must be checked.
```
replication of deletions or expirations are not prevented...

tonyjhillman replied:
I've added. I've also added some information to the front of the subsection in xdcr-filtering.adoc, explaining the function of each of the filters in a little more detail.

nelio2k commented on modules/learn/pages/clusters-and-availability/xdcr-filtering.adoc:
```diff
@@ -163,3 +170,72 @@ image::xdcr/filter-replication-diagram-7.png[,720,align=left]
+
+Either of the following approaches can be used to ensure that no migrated data is lost:
+
+* Configure deletion filters to prohibit the replication of deletions.
```
deletions or expirations

tonyjhillman replied:
I've now tried to add "expirations, and/or TTLs" where appropriate, throughout.

nelio2k commented on modules/manage/pages/manage-xdcr/replicate-using-scopes-and-collections.adoc:
```diff
@@ -262,6 +275,20 @@ Note the confirmatory notification that appears underneath the replication-defin
++
+image::manage-xdcr/filter-xdcr-deletion-filters.png[,320,align=left]
++
+For each filter, to ensure that deletion is replicated, leave the setting at its default (with the checkbox unchecked); and to ensure that deletion is _not_ replicated, check the checkbox.
```
or expiry
(check the expiry checkbox)

nelio2k replied:
(Don't see deletion filter in XDCR adv filtering page)

If a source document contains an expiry, and the the user does not want the expiry to apply to the document when it reaches the target bucket, check the "TTL" checkbox...
This will cause the document's expiry to be set to 0 prior to transmission to the target.
However, this also means that the target bucket's expiry setting, if exists, will take effect.

tonyjhillman replied:
I've included mention of expirations. As mentioned above, I've added an introduction that provides info on each deletion filter, including TTL.

nelio2k reviewed (commented):
Added a few in-line comments

tonyjhillman committed:
DOC-8558: Rewording

tonyjhillman committed:
DOC-8558: New introduction to deletion filters section

tonyjhillman committed:
DOC-8588: Additional information on deletion filters

tonyjhillman committed:
DOC-8558: Rewording in various passages

tonyjhillman commented:
I've added as per the inline comments. I've also added, on the main advanced filtering page, an intro to deletion filters that describes each activity; and also passages on Tombstones and Replication, Expiration, TTL and Replication, and Stale-Data Replication.

nelio2k commented on modules/learn/pages/clusters-and-availability/xdcr-filtering.adoc:
```diff
@@ -174,18 +174,69 @@ A successful match is made, and `airline_8838` is duly replicated to the target
+
+When a document is deleted or is expired, a tombstone is created.
+Tombstones and their management are described in xref:learn:buckets-memory-and-storage/storage.adoc#tombstones[Tombstones].
+In order to replicate a deletion or an expiration, XDCR must be able to find, on the source, a tombstone that corresponds to the deleted or expired document.
```
XDCR doesn't really "find" it...
XDCR, as a recipient of the source KV DCP,  must be able to receive a deletion from the DCP stream

Not sure how to best explain this concept without confusing the user with technicalities...

tonyjhillman replied:
I've modified this to include the notion of DCP.

nelio2k commented on modules/learn/pages/clusters-and-availability/xdcr-filtering.adoc:
```diff
@@ -174,18 +174,69 @@ A successful match is made, and `airline_8838` is duly replicated to the target
+When a document is deleted or is expired, a tombstone is created.
+Tombstones and their management are described in xref:learn:buckets-memory-and-storage/storage.adoc#tombstones[Tombstones].
+In order to replicate a deletion or an expiration, XDCR must be able to find, on the source, a tombstone that corresponds to the deleted or expired document.
+When the tombstone is located, XDCR generates a corresponding deletion or expiration event, and replicates this to the target.
```
located -> received (from source DCP)

tonyjhillman replied:
Here also, I've added the notion of DCP.

nelio2k commented on modules/learn/pages/clusters-and-availability/xdcr-filtering.adoc:
```diff
@@ -174,18 +174,69 @@ A successful match is made, and `airline_8838` is duly replicated to the target
+If a document has been deleted or expired, and the resulting tombstone has been purged prior to XDCR being able to locate it, no deletion or replication event is replicated.
+This situation might occur if:
+
+* A document is deleted and then immediately recreated, such that the time during which the tombstone existed has been too brief for location of the tombstone by XDCR to occur.
```
The tombstone technically still exists... however, the KV engine could decide that the tombstone is superceded by the recreated document, and for the sake of efficiency (deduplication), skip sending the deletion and just send the recreated document instead.

The phrase "tombstone existed has been too brief" would be incorrect and runs contrary to the concept of tombstone purging.

tonyjhillman replied:
I've rewritten this accordingly.

tonyjhillman committed:
DOC-8558: Modifications to tombstones-related passages
